dot/state/block_finalisation.go

```diff
@@ -176,6 +171,21 @@ func (bs *BlockState) SetFinalisedHash(hash common.Hash, round, setID uint64) er
 		),
 	)
 
+	if !bs.lastFinalised.Equal(hash) {
+		defer func(lastFinalised common.Hash) {
```

Comment:
Why defer? Can we just have

```go
previousLastFinalised := bs.lastFinalised
bs.lastFinalised = hash
if !previousLastFinalised.Equal(hash) {
    // defer function body, replacing lastFinalised with previousLastFinalised
}
return nil
```

Reply:
I chose a defer, cause in normal operation the in memory tries should always contain the last finalised trie.  But if we return an error here then a lot of tests will fail outside the `dot/state` package.  I'd rather not just ignore the error since every other error in this function isn't ignored.

Reply:
I didn't suggest to return an error, just have it without the defer function.

Reply:
yea.. I dislike ignoring errors, and I feel like cleaning up the in-memory tries is something that can be encompassed in a defer after the finalised block hash is updated.  If we end up refactoring some of this code, I think this should return an error since we should be ensuring the integrity of what's in-memory and what's on disk.

Reply:
If you want to keep the defer statement, at least make it a separate function?
It's a bit confusing/ugly as it is now I think